### PR TITLE
Improve error message when not providing a value for JSX key

### LINKED ISSUE #12931: (plugin-transform-react-jsx): Compile error, when element's JSXAttribute key isn't given a value
## Bug Report

- [x] I would like to work on a fix!

**Current behavior**

When compiling a React component that returns an array of elements and we missuse the key prop, we only get the following error:
```
TypeError: <filename>: Property arguments[2] of CallExpression expected node to be of a type ["Expression","SpreadElement","JSXNamespacedName","ArgumentPlaceholder"] but instead got undefined
```

**Input Code**

```js
function Component() {
  return (
    [
      <div key></div>
    ]
  )
}

```

**Expected behavior**

I want the console error to be more helpful on why the build process failed or the build to be successful.

**Babel Configuration**
- CRA v4.0.2
https://github.com/facebook/create-react-app/tree/v4.0.2/packages/babel-preset-react-app

**Environment**

```
 System:
    OS: Linux 5.4 Ubuntu 20.04.2 LTS (Focal Fossa)
  Binaries:
    Node: 12.20.1 - ~/.nvm/versions/node/v12.20.1/bin/node
    Yarn: 1.22.5 - /usr/bin/yarn
    npm: 6.14.10 - ~/.nvm/versions/node/v12.20.1/bin/npm
  npmPackages:
    babel-eslint: ^10.0.3 => 10.1.0 

```
- Babel version(s): v7.12.3
- @babel/plugin-transform-react-jsx" v7.12.17
- Monorepo: no

**Possible Solution**

A warning should be printed after the transformed value is empty(null) for the attribute named "key". Or maybe the true boolean literal should be applied, as stated in the documentation for JSX props.

https://github.com/babel/babel/blob/265424d43f964cafd976a2143ddeee4066e14fd6/packages/babel-plugin-transform-react-jsx/src/create-plugin.js#L406-L423

**Additional context**

I don't think this is a big issue since key values are supposed to be unique in every list. It would save some debugging for others, especially since it doesn't occur in development mode. I'm ready and happy to help improve the code.

### PR DESCRIPTION
**Pull request by hajnalbendeguz**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12931 
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR Link    | - 
| Any Dependency Changes?  | None
| License                  | MIT

Throwing a more descriptive error when compiling a React element in a list that hasn't been given a value for its key.
eg. `<Component key />`

### CODE REVIEW

File: packages/babel-plugin-transform-react-jsx/src/create-plugin.js [hunk 1]
```diff
@@ -411,9 +411,17 @@ You can set \`throwIfNamespace: false\` to bypass this warning.`,
             case "__self":
               if (extracted[name]) throw sourceSelfError(path, name);
             /* falls through */
-            case "key":
-              extracted[name] = convertAttributeValue(attr.node.value);
+            case "key": {
+              const keyValue = convertAttributeValue(attr.node.value);
+              if (keyValue === null) {
+                throw path.buildCodeFrameError(
+                  'Provide an explicit value to be used as a key. You are not supposed to use the "key" shorthand for "key={true}".',
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-react-jsx/src/create-plugin.js:418]: I'm not a native speaker so I might just be reading this wrong, but this message _feels_ a bit aggressive to me.

Maybe this?

```suggestion
                  'Please provide an explicit key value. Using "key" as a shorthand for "key={true}" is not allowed.',
```

Comment (existentialism) [reply packages/babel-plugin-transform-react-jsx/src/create-plugin.js:418]: Agreed

Comment (hajnalbendeguz) [reply packages/babel-plugin-transform-react-jsx/src/create-plugin.js:418]: Okay, this seems nicer. I've included the change in a seperate commit.

File: packages/babel-plugin-transform-react-jsx/test/fixtures/react-automatic/should-disallow-valueless-key/input.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+
```
[the remark this answers is not in the document]
Comment (hajnalbendeguz) [reply packages/babel-plugin-transform-react-jsx/test/fixtures/react-automatic/should-disallow-valueless-key/input.js:1]: Okay. I've done that. As I mentioned in the issue, the compilation was successful in development mode. Fixing it was a bit trickier. Let me know if I have to improve my solution in any way.

**Review by existentialism — APPROVED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]
We should also throw in development mode, but let's merge this as-is for now since it already fixes the original problem (the unhelpful error message).

If you want to open the follow-up PR please do so, otherwise I'll open an issue about it!

**Comment by yedidyak** [comment]
This changed the behaviour very slightly, and actually ended up breaking a build of ours. Now when running the react-native metro bundler on an error case of 'key' used as a boolean it throws an exception and fails at compile time, instead of at runtime.

We fixed it by fixing the bug, but maybe worth checking this scenario in future.

Thanks for working on amazing open source anyway!

**Comment by nicolo-ribaudo** [comment]
Oh sorry for the inconvenience, I assumed this was already a compile-time error (just with a worse message) because of https://github.com/babel/babel/issues/12931.

If many people are affected we could make it a warning instead, but since this would be a runtime error anyway I lean towards leaving it as-is.

**Pull request merged**